File: packages/blitz/src/utils/crypto.ts
/** i32 * i32 */
const imul = Math.imul;

function cyrb53(str: string, seed = 0) {
  let h1 = 0xdeadbeef ^ seed;
  let h2 = 0x41c6ce57 ^ seed;
  for (let i = 0, ch; i < str.length; i++) {
    ch = str.charCodeAt(i);
    h1 = imul(h1 ^ ch, 0x85ebca6b);
    h2 = imul(h2 ^ ch, 0xc2b2ae35);
    h1 = (h1 << 13) | (h1 >>> 19);
    h2 = (h2 << 15) | (h2 >>> 17);
  }

  h1 = imul(h1 ^ (h1 >>> 16), 0x85ebca6b) ^ imul(h2 ^ (h2 >>> 13), 0xc2b2ae35);
  h2 = imul(h2 ^ (h2 >>> 16), 0x85ebca6b) ^ imul(h1 ^ (h1 >>> 13), 0xc2b2ae35);

  return [h1 >>> 0, h2 >>> 0];
}

function toBase64(array: Uint8Array) {
  const string = String.fromCharCode.apply(null, Array.from(array));
  return btoa(string);
}

export function hashRef(message: string) {
  const [x, y] = cyrb53(message);
  const buffer = new ArrayBuffer(8);
  {
    const dataview = new DataView(buffer);
    dataview.setUint32(0, x);
    dataview.setUint32(4, y);
  }
  return toBase64(new Uint8Array(buffer)).slice(0, 7);
}

File: packages/swc-plugin-remove-server-code/src/crypto.rs
use base64::{engine::general_purpose::STANDARD, Engine};

fn cyrb53(s: &str, seed: u32) -> (u32, u32) {
    let mut h1: u32 = 0xdeadbeef ^ seed;
    let mut h2: u32 = 0x41c6ce57 ^ seed;

    for &ch in s.as_bytes() {
        h1 = (h1 ^ (ch as u32)).wrapping_mul(0x85ebca6b);
        h2 = (h2 ^ (ch as u32)).wrapping_mul(0xc2b2ae35);
        h1 = h1.rotate_right(19);
        h2 = h2.rotate_right(17);
    }

    h1 = (h1 ^ (h1 >> 16)).wrapping_mul(0x85ebca6b) ^ (h2 ^ (h2 >> 13)).wrapping_mul(0xc2b2ae35);
    h2 = (h2 ^ (h2 >> 16)).wrapping_mul(0x85ebca6b) ^ (h1 ^ (h1 >> 13)).wrapping_mul(0xc2b2ae35);

    (h1, h2)
}

fn to_base64(bytes: &[u8]) -> String {
    STANDARD.encode(bytes)
}

pub fn hash_ref(message: &str) -> String {
    let (x, y) = cyrb53(message, 0);
    let mut buffer = [0u8; 8];
    buffer[..4].copy_from_slice(&x.to_be_bytes());
    buffer[4..].copy_from_slice(&y.to_be_bytes());
    to_base64(&buffer)[..7].to_string()
}
